Adds CONTRIBUTING.md

tansaku opened the pull request:
fixes #799 

This is a CONTRIBUTING.md doc based on the one we have for LocalSupport.  GitHub automatically links this file into the UI for pull requests and many developers will expect to find help and advice on how to contribute in this file:

https://help.github.com/articles/setting-guidelines-for-repository-contributors/

This is a proposal based on the approximate workflow I've been following since we switched to waffle, and I'm very interested to hear about how we can word this better, make changes etc. so that we have a workflow that works well for as many of us as possible.

tansaku commented:
ah and we should cross check this against existing documentation like http://www.agileventures.org/projects/websiteone/documents/how-to-submit-a-pull-request-on-github and others that I might not have found yet

tansaku commented:
and https://github.com/AgileVentures/WebsiteOne/wiki/Project-coordination-outline

tochman commented:
Let's set up a chore (or however you are labeling this in Waffle) to migrate all legacy documents that are still relevant from WSO and place it in the repository if that is the flow that is agreed on.

tansaku commented:
great suggestion @tochman 

have created that at #805

tansaku commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,37 @@
+```
+
+which will associate the pull request with the issue in the Waffle board.
+
````
Michael says that we should also note to resync to develop before making pull request

tansaku commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,37 @@
+```
+
+Please ensure that each commit in your pull request makes a single coherent change and that the overall pull request only includes commits related to the specific GitHub issue that the pull request is addressing.  This helps the project managers understand the PRs and merge them more quickly.
+
````
Michael says that we should also note to resync to develop before making pull request

tochman commented:
I propose that updating the README is included in this PR. That would allow for pointing to the CONTRIBUTING.md from README and avoid confusion. Also, the project description on AgileVentures should be updated with information about the new workflow (at least remove the reference to PT).

tansaku commented:
@tochman I agree we should update the README, but I would say let's do that in another pull request with another issue.  We've got a separate issue for that #811 

Let's make sure that we are happy with the approximate wording in this document and get it merged in.  The README update can follow in a subsequent PR very quickly.

To me the absolutely critical thing from the point of view of successful project managment is to get into the habit of working around micro PRs and merging them fast

sampritipanda commented:
@tansaku Issue 811 has a task for updating the Contributing Section.

I think updating the readme and contributing file are related and would be better if they were done in one PR.

tansaku commented:
@sampritipanda in other situations I might agree, but I think that sometimes we need to go to the extreme to explore things, and make sure everyone is on the same page about how a process operates.  That is something that has worked really well in LocalSupport in terms of managing the project and achieving manageable pull requests.

Regarding #811 having a section on fixing the contribution section, that seems a little out of band for that ticket, since we already have the open issue #799 for updating the contribution document and we have a pull request specifically for that.

My preference is very very very strongly that we comment on the specific engineering flow in the CONTRIBUTIONS.md in this pull request.  I think it's very important we stay task focused on that - get agreement on that, and get that merged in.  I think effective process in that regard is very important for the long term manageability of this project.  If everyone is very happy with the current state of CONTRIBUTIONS.md and that's why we've moved onto discussing the README, then that's great, but I just want to double check that before merging in.

That said, for the time I could have been writing the above I could have been editing the README, but the fact that I wrote the above is a reflection of just how much importance I place on staying task focused *and* that people are happy with what's actually in CONTRIBUTING.md, and ensuring that pull requests and commits are as tightly focused as possible.

I think adding the README changes to this PR is dangerous feature creep and should be carefully avoided.

tochman commented:
The way I see it, the purpose of adding this document is to bring more clarity for new users, so in that regard, not only adding/updating README it is connected to this PR but also the task of migrating the rest of the project documentation to this repo. The CONTRIBUTING document in itself will not achieve the overall goal.

I have no objections as to the content of the document itself. I covers parts of the new workflow that has been put in place and thus suits its purpose. I might have objections to the workflow itself, but that is outside the scope of this PR.

diraulo commented:
btw @tansaku by fixing the Contributing section, I was referring to the one present in the README.md file. Either by removing it or update it to point to the new document CONTRIBUTING.md

diraulo commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,44 @@
+Code Style
+-------------
+
+We recommend the [Ruby Style Guide](https://github.com/bbatsov/ruby-style-guide)
```
@tansaku should we setup houndci to check coding style?

tansaku replied:
@diraulo that's something we now have on LocalSupport - it's something I'd like to add at some point, but I'm not in a rush.  Depending on how it's configured it can come across as bothersome for some developers ...

tochman commented:
Apart from other things I would propose having a section about who's responsibility it is to do code reviews and merge PR's and about CI and deployment. I also notice that test coverage (coveralls.io) is not measured on PR's in this project (anymore). Perhaps we should reintroduce that and have a section about tests and that we don't like when coverage decreases on PR's. These things can clarify the process for new contributors.

tochman commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,44 @@
+We use [Waffle](https://waffle.io/AgileVentures/WebsiteOne) to manage our work on features, chores and bugfixes.  Every pull request should a corresponding GitHub issue, and when you create feature/bug-fix branches please include the id of the relevant issue, e.g.
+
+```
+git checkout -b 799_add_contributing_md
````
Is this a Waffle requirement? Looks messy to have to reference the issue number in the branch name.

tansaku replied:
It's a waffle best practice.  I've been using it consistently for a year and I love it.  My autocomplete searching for branches is faster - I always know which issue I'm referring to ...

* https://github.com/waffleio/waffle.io/wiki/Recommended-Workflow-Using-Pull-Requests-&-Automatic-Work-Tracking

* https://github.com/waffleio/waffle.io/wiki/FAQs

tochman commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,44 @@
+Our default working branch is `develop`.  We do work by creating branches off `develop` for new features and bugfixes.  Any feature should include appropriate Cucumber acceptance tests and RSpec unit tests.  We try to avoid view and controller specs, and focus purely on unit tests at the model and service level where possible.  A bugfix may include an acceptance test depending on where the bug oc ...
+
+Each developer will usually work with a [fork](https://help.github.com/articles/fork-a-repo/) of the [main repository on Agile Ventures](https://github.com/AgileVentures/WebSiteOne). Before starting work on a new feature or bugfix, please ensure you have [synced your fork to upstream/develop](https://help.github.com/articles/syncing-a-fork/):
+
```
I think that we need to ensure that new members understand what `upstream` is and how you set that remote. (I believe we had a section on that in a document on WSO. (http://www.agileventures.org/projects/websiteone/documents/how-to-submit-a-pull-request-on-github)

```
Assuming that you have the following local setup:
$ git remote -v
origin	   git@github.com:YourName/WebsiteOne.git (fetch)
origin	   git@github.com:YourName/WebsiteOne.git (push)
upstream   https://github.com/AgileVentures/WebsiteOne.git (fetch)
upstream   https://github.com/AgileVentures/WebsiteOne.git (push)
```

tansaku replied:
good call - I wonder if we should pull that into a docs section ....

tochman commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,44 @@
+
+Please ensure that each commit in your pull request makes a single coherent change and that the overall pull request only includes commits related to the specific GitHub issue that the pull request is addressing.  This helps the project managers understand the PRs and merge them more quickly.
+
+Whatever you are working on, or however far you get please do open a "Work in Progress" (WIP) [pull request](https://help.github.com/articles/creating-a-pull-request/) so that others in the team can comment on your approach.  Even if you hate your horrible code :-) please throw it up there and we'll help guide your code to fit in with the rest of the project.
```
About setting WIP status on a PR. In the PR title or as a badge?

tansaku replied:
only those with write access to the repo can set it as a label, so for new contributors it would be like "[WIP] my awesome contribution", which is what we do on LocalSupport

tochman commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,44 @@
+fixes #799
+```
+
+which will associate the pull request with the issue in the Waffle board.
````
Could we also ask for some description on what changes has been made?

tansaku replied:
absolutely, and other things such as screenshots or deployed version of site - but we want to get the balance right I guess ...

tansaku commented:
@tochman said 

> Apart from other things I would propose having a section about who's responsibility it is to do code reviews and merge PR's and about CI and deployment. I also notice that test coverage (coveralls.io) is not measured on PR's in this project (anymore). Perhaps we should reintroduce that and have a section about tests and that we don't like when coverage decreases on PR's. These things can clarify the process for new contributors.

All agreed - After lunch I'll add a section on responsibilities (unless someone else wants to add one first), and here's a ticket on getting code coverage measured ... #814

tansaku commented:
@tochman said

> The way I see it, the purpose of adding this document is to bring more clarity for new users, so in that regard, not only adding/updating README it is connected to this PR but also the task of migrating the rest of the project documentation to this repo. The CONTRIBUTING document in itself will not achieve the overall goal.

Understood.  The primary goal for me with this issue (i.e. #799) is to get general consensus between @tochman @tansaku @sampritipanda @diraulo and @dalupus about our engineering flow.  I say those five names as these are the five who have write access and have been active in the last month.  Of course I am very interested in input from others, but this group seems key at the moment.

Perhaps the PR should have gone into an ENGINEERING.md instead and I have muddied the water by dropping it into CONTRIBUTING.md  - and for that my apologies. 

I do ultimately think that CONTRIBUTING.md is the place for this document, and we have another PR for changing the README and that can me merged in parallel (so it's not big odds either way?) Migrating all the old docs is another important task which we need to start, but I think both of those are lesser issues compared to achieving approximate consensus on the engineering flow.

@tochman also said

> I have no objections as to the content of the document itself. I covers parts of the new workflow that has been put in place and thus suits its purpose. I might have objections to the workflow itself, but that is outside the scope of this PR.

At the risk of opening a can of worms :-) I think objections to the workflow are in scope for this PR.  We won't necessarily resolve them, but it might be good to hear what they are.

I think we should always be all ears to hear about ways that we can optimize our process, although I'm very happy for that to be brought up in the retrospective if you think that's a better forum?

tochman commented:
Here comes the 'can of worms' ;-)

Yes, so I'm not in found of the recent changes as to what tools we are using (Waffle) and the impact it has on the workflow (like the forced naming of the branches or PR's). I'm used to another workflow that also asked for certain standards (like links to PT story, user stories for all features, a verbose description of the PR, etc). 

I also fell a little intimidated by references to best practices. It is one thing to propose best practices, another to impose those practices on devs. Even if it is always good to learn new things. Also, my personal opinion is that the workflow actually got more complicated due to those changes and makes it harder to onboard new developers. But that is just a theory. I get lost in it, at least. 

I totally agree that PR's and commit should be clear, focused and precise (God knows I wish that had been the case when I was leading this project). But let's be frank, it does not matter how well we document things and how much resources we spend on simplifying the on-boarding process, we will still face a lot of confusion and questions from new members - questions that are best answered á la minute in person, like in a standup. And best practices, as any other knowledge, are best received and put into use when they actually make sense. 

I agree that all projects needs a reboot at some point but I feel that abandoning PT in this case feels like building a new kitchen when there is a lot of dirty dishes in the sink. You know, instead of actually making the dishes. What I'm saying is that we could have cleaned up a lot of stories in PT and focused on delivering features. The problem of this project was never the workflow in itself but rather lack of leadership in recent times. We've known for at least a year that we need to have a devoted project manager for this project to succeed. 

Also, the lack of a client is burdensome. For better and for worse, when we had a designated client we could have an discussion about accepting and rejecting features. Now I feel that we are sailing without a direction. 

A final reflection, I am worried that the work with the process and the form itself is taking resources and time from actual work. I waited more than a week for an actual code review (that exposed flaws on on of my PR's) and there are Issues and feature requests on Waffle that has not been addressed, discussed, assessed and started on. I'd say let's focus on getting shit done and ship some code ;-)

Yeah, so.. sorry about this.... I'd love to discuss more in the retro, but I tend to forget things so I wanted to write this down now. (It's an age thing. also, sorry for the bad spelling. I have no time to do a spellcheck)

tansaku commented:
thanks @tochman really appreciate you sharing your honest opinion there - very valuable

tansaku commented:
I've updated the pull request based on the feedback from @tochman and others and added a note to http://www.agileventures.org/projects/websiteone/documents/wso-project-workflow about the documentation review process that we are going through.

Unless anyone has any blockers on the current document I would propose to merge both this pull request and #813 tomorrow and make subsequent changes by further pull request
